### PR TITLE
Retry when database is locked

### PR DESCRIPTION
**Pull request by hallyn**

### CODE REVIEW

File: lxd/db.go [hunk 1]
```diff
@@ -402,14 +403,21 @@ func dbImageGet(d *Daemon, name string, public bool) (*shared.ImageBaseInfo, err
...
-		return nil, err
+	for {
+		err = d.db.QueryRow(q, (name + "%")).Scan(&countImg)
+		if err == nil {
+			break
+		} else if ! shared.IsDbLockedError(err) {
+			shared.Debugf("dbImageGet: search for count of image by fp return error: %q\n", err)
+			return nil, err
+		}
+		time.Sleep(100 * time.Millisecond)
```
Comment (stgraber) [inline lxd/db.go]: I think we ought to have a timeout at some point, in the unlikely event that we've deadlocked ourselves.

Can we return an error after a 30s wait?

File: lxd/db.go [hunk 2]
```diff
@@ -422,7 +430,17 @@ func dbImageGet(d *Daemon, name string, public bool) (*shared.ImageBaseInfo, err
...
+	for {
+		err = d.db.QueryRow(q, (name+"%")).Scan(&image.Id, &image.Fingerprint, &image.Filename, &image.Size, &image.Public)
+		if err == nil {
+			break
+		} else if ! shared.IsDbLockedError(err) {
+			shared.Debugf("dbImageGet: search for image by fp returned error %q\n", err)
+			return nil, err
+		}
+		time.Sleep(100 * time.Millisecond)
+	}
```
Comment (stgraber) [inline lxd/db.go]: Same wrt, deadlock timeout.

File: lxd/db.go (hunk 1 above)
Comment (stgraber) [inline lxd/db.go]: Also why can't we use a wrapper here?

File: lxd/db.go (hunk 2 above)
Comment (stgraber) [inline lxd/db.go]: And same question, wrt wrapper.

File: lxd/db.go [hunk 3]
```diff
@@ -421,7 +424,13 @@ func dbImageGet(d *Daemon, name string, public bool) (*shared.ImageBaseInfo, err
 	if public {
 		q = q + " AND public=1"
 	}
-	err = d.db.QueryRow(q, (name+"%")).Scan(&image.Id, &image.Fingerprint, &image.Filename, &image.Size, &image.Public)
+
+	arg2 = []interface{}{&image.Id, &image.Fingerprint, &image.Filename, &image.Size, &image.Public}
+	err = shared.DbQueryRowScan(d.db, q, arg1, arg2)
+	err = shared.DbQueryRowScan(d.db, q, arg1, arg2)
```
Comment (stgraber) [inline lxd/db.go]: Hmm, what?

Comment (hallyn) [inline lxd/db.go]: On Mon, Mar 16, 2015 at 08:51:01PM -0700, St�phane Graber wrote:
> > @@ -421,7 +424,13 @@ func dbImageGet(d *Daemon, name string, public bool) (*shared.ImageBaseInfo, err
> >  	if public {
> >  		q = q + " AND public=1"
> >  	}
> > -	err = d.db.QueryRow(q, (name+"%")).Scan(&image.Id, &image.Fingerprint, &image.Filename, &image.Size, &image.Public)
> > +
> > +	arg2 = []interface{}{&image.Id, &image.Fingerprint, &image.Filename, &image.Size, &image.Public}
> > +	err = shared.DbQueryRowScan(d.db, q, arg1, arg2)
> > +	err = shared.DbQueryRowScan(d.db, q, arg1, arg2)
> 
> Hmm, what?

Some people like to shout, I like to calmly repeat myself.

**Pull request merged**